### DISCUSSION TITLE
Customize Background

### DISCUSSION THREAD

**Discussion by Dimonka2** [bcakmakoglu/vue-flow#37] [Q&A] [answered]
Hi @bcakmakoglu,
"Background" component is very useful and works well. 
I am wondering is there a functionality in a Background component to implement some custom painting? Like paint color regions with text (e.g. Level1/Level2/Level3).

Currently I am thinking to create some custom annotation readonly nodes, but maybe there are other ways to draw additional information.
Thanks!

**Comment by bcakmakoglu** [comment, marked as answer]
Hi.

Currently that wouldn't work; The Background stretches over the whole Flow with a singular pattern.
I am guessing you want something like this, right?
![grafik](https://user-images.githubusercontent.com/78412429/158164475-2ae1171d-bb31-4bde-b03c-9acaa1a0fa4f.png)

With a couple of small changes that would be possible and I could add those changes to 0.4.0.
Usage would look something like this 
```vue
<VueFlow v-model="elements" class="vue-flow-basic-example">
    <Background :variant="BackgroundVariant.None" :height="33" bg-color="pink">
      <text fill="#000000" font-size="22" font-family="ARIAL" x="120" y="110"> LEVEL 1 </text>
    </Background>
    <Background :height="33" bg-color="purple">
      <text fill="white" font-size="22" font-family="ARIAL" x="120" y="110"> LEVEL 2 </text>
    </Background>
    <Background :height="33" bg-color="crimson">
      <text fill="white" font-size="22" font-family="ARIAL" x="120" y="110"> LEVEL 3 </text>
    </Background>
    <MiniMap />
    <Controls />
  </VueFlow>
```

**Reply by bcakmakoglu** [reply]
Fyi: If you're trying to achieve nested nodes, that's a feature that's coming with 0.4.0 too.
![grafik](https://user-images.githubusercontent.com/78412429/158165145-eb4a8da2-69c9-4715-93ed-52c0cd4607d2.png)

**Reply by bcakmakoglu** [reply]
I will probably deploy the Background changes with the next pre-release which should be today (latest tomorrow), i.e. you can install them by using the `@next`-tag (`npm i @braks/vue-flow@next`)

**Comment by Dimonka2** [comment]
Thank you! 
Actually the option # 1 is exactly what I was hoping for. So it would be great if you will be able to implement something like this in the following updates!

**Reply by bcakmakoglu** [reply]
I've added the suggested changes to the newest release, so if you're using 0.4.0-7 you should be able to replicate the example above

**Comment by Dimonka2** [comment]
> I've added the suggested changes to the newest release, so if you're using 0.4.0-7 you should be able to replicate the example above

This feature is really cool, but the issue is that the background is not panned together with nodes.. It stays fixed, but ideally it should be moved together with nodes.

**Reply by bcakmakoglu** [reply]
You can offset the background position with the position of a node:
```vue
<Background
      :variant="BackgroundVariant.None"
      :height="33"
      bg-color="pink"
      :x="elements[0].position.x"
      :y="elements[0].position.y"
    >
```

But that would just cause backgrounds to overlap/cut each other.
Like this:
![grafik](https://user-images.githubusercontent.com/78412429/158840676-2a4e6722-63d4-4468-967c-b84d9c1ee396.png)

What you describe sounds more like nested nodes now; Where the parent-node is just a big area in which nodes can be dropped to and the area moved (i.e. all nodes moved at once).

Either way technically you can already implement a "tracking" background or a parent node with enough area to cover your nodes.
You can get the dimensions of a box of nodes with the `getRectOfNodes` utility function.
Pass the nodes you want your box to wrap around and it will return a `Rect` i.e. width, height, x, y which you can then use for your background/node.
You might want to scale those values with the current zoom value, which is stored as `transform.value[2]`.
